### PR TITLE
feat(discord.js-utilities): add checking index changes and documentation

### PR DESCRIPTION
**Pull request by RealShadowNova**
This PR implements a feature `disableResponseEdit` to the `IPaginatedMessageAction`. This enables developers to have an action only execute without editing the page to the current index.

This PR also includes documentation for the utility as a whole. Some of which was not written the best, so if you think something should be different go right ahead and suggest it.

### CODE REVIEW

**Review by Favna — DISMISSED** [review]

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts [hunk 1]
```diff
@@ -1,65 +1,162 @@
...
 	public index = 0;
 
+	/**
+	 * The amount of time to idle before the paginator is closed. Defaults to `20 * 1000`.
+	 */
 	public idle = 20 * 1000;
 
+	/**
+	 * Constructor for the [[PaginatedMessage]] class
+	 * @param __namedParameters The [[PaginatedMessageOptions]] for this instance of the [[PaginatedMessage]] class
```
Comment (kyranet) [inline packages/discord.js-utilities/src/lib/PaginatedMessage.ts:70]:
```suggestion
	 * @param options The [[PaginatedMessageOptions]] for this instance of the [[PaginatedMessage]] class
```

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts [hunk 2]
```diff
@@ -93,30 +190,43 @@ export class PaginatedMessage {
 
 				const page = await this.resolvePage();
 
-				await response.edit(page!);
+				if (!action.disabledResponseEdit) await response.edit(page!);
```
Comment (kyranet) [inline packages/discord.js-utilities/src/lib/PaginatedMessage.ts:193]: Shouldn't the resolveImage be skipped alongside the response.edit?

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts [hunk 3]
```diff
@@ -170,11 +281,39 @@ export class PaginatedMessage {
...
+ *   run: ({ response, collector }) => {
+ *     await response!.reactions.removeAll();
+ *     collector!.stop();
+ *   }
+ * }```
+ */
 export interface IPaginatedMessageAction {
 	id: string;
+	disabledResponseEdit?: boolean;
 	run(context: PaginatedMessageActionContext): Awaited<unknown>;
```
Comment (kyranet) [inline packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: To be honest, I'm not a fan of doing this statically, there are two solutions:

- **Semver Minor**: We track the index state, if it changed, after the call, we edit the message. This option is more intuitive and easier to use.
- **Semver Major**: Make the function accept a value that tells whether or not the message should be refreshed.

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts (hunk 1 above)
Comment (Favna) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:70]: Wrong. TypeDoc named destructured parameters as `__namedParameters`. Using `options` here will cause a mismatch.

### With `options`:
***There is no description for the options***
image


### With `__namedParameters`:
***There is a description for the options***
image

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts (hunk 3 above)
Comment (RealShadowNova) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: Ok, I get that I can change that right now.
I could probably implement both tbh.
Would the action's run method return a boolean whether to edit the response or not? If that's the case we would explicitly check the type of the return value to see if it's boolean instead of converting it to a boolean.

Comment (kyranet) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: If you implement both, you'll end up with 2.0.0 instead of 1.1.0 if you implement only once.
Besides, they conflict with each other.

Comment (RealShadowNova) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: Then can't we handle that conflict? A couple conditions on whether the response should be edited:

- Action's boolean is `true` AND index hasn't changed THEN do not edit response.
- Action's boolean is `false` AND index hasn't changed THEN edit response.
- Action's boolean is `true` AND index has changed THEN do not edit response.
- Action's boolean is `false` AND index has changed THEN edit response.
- Action's boolean is invalid AND index hasn't changed THEN do not edit response.
- Action's boolean is invalid AND index has changed THEN edit response.

In other words if the boolean is `true` then do NOT edit the response, otherwise edit.

Comment (kyranet) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: Please don't try to make this API any harder than my Intelligent Behaviour System's practices, keep it simple, just edit when index was changed.

Comment (RealShadowNova) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:311]: Alright, I'll change that.

**Review by kyranet — APPROVED** [review]

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts [hunk 4]
```diff
@@ -1,65 +1,162 @@
...
+ * @example
+ * ```typescript
+ * const handler = new PaginatedMessage();
+ * ```
+ *
+ * @example
+ * ```typescript
+ * // To utilize actions you can use the IPaginatedMessageAction by implementing it into a class.
+ *
+ * class ForwardAction implements IPaginatedMessageAction {
```
Comment (vladfrangu) [inline packages/discord.js-utilities/src/lib/PaginatedMessage.ts:22]: This presents a class...however, said class would need to be instantiated *somewhere and by someone*

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts (hunk 4 above)
Comment (RealShadowNova) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:22]: Yes, that's correct which is why if you read further in that same code block it states it can also pass an object.

**Review by Favna — APPROVED** [review]

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts (hunk 4 above)
Comment (vladfrangu) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:22]: Doesn't change the fact you're presenting the possibility of people using raw classes in addAction, which is not handled. Either document that you NEED to pass an instance yourself, or handle it in the code

Comment (RealShadowNova) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:22]: What do you think `IPaginatedMessageAction` is? It's an interface, meaning you can implement it into a class or directly an object. It clearly states this in the documentation. Your reasoning does not make sense imo.

Comment (kyranet) [reply packages/discord.js-utilities/src/lib/PaginatedMessage.ts:22]: `class X implements IPaginatedMessageAction {}` only means that instances of `X` are assignable to the type `IPaginatedMessageAction`, not `X` itself.

File: packages/discord.js-utilities/src/lib/PaginatedMessage.ts [hunk 5]
```diff
@@ -1,65 +1,163 @@
...
+ *
+ * @example
+ * ```typescript
+ * const handler = new PaginatedMessage();
+ * ```
+ *
+ * @example
+ * ```typescript
+ * // To utilize actions you can use the IPaginatedMessageAction by implementing it into a class.
+ * // [[PaginatedMessage]] requires you to have the class initalized.
```
Comment (kyranet) [inline packages/discord.js-utilities/src/lib/PaginatedMessage.ts:21]:
```suggestion
 * // [[PaginatedMessage]] requires you to have the class initialized using `new`.
```

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

**Review by vladfrangu — APPROVED** [review]

**Review by Favna — APPROVED** [review]
